Redbridge Harvester

kjoshi opened the pull request:
# Redbridge Harvester
This is very similar to the other harvesters. The Redbridge API doesn't have a single "get all datasets" API endpoint, so the process of finding all the available datasets is a bit convoluted. But once the package_dicts are created the import process is pretty standard.

I've moved a few things out of the `datapress` and `nomis` harvesters and into the `util` module to try and make it easier to do the things that need doing for most harvesters. There's some common functionality getting spread around the different harvesters that it would be good to try and have in one place. I'm not sure if `util` is the best place but hopefully it'll do for now.

There are a couple of `util` things that I think might need to be added to the Camden harvester:
 - The bit that checks for datasets that have been deleted upstream and removes them from the local CKAN (e.g. lines 145-149 of `redbridge.py` and some bits in the import stage)
 - The bit that transfers across any existing package_extras (e.g, line 231 of `redbridge.py`)

madelinekosse committed:
DAT-409 - Add Socrata harvester

Adds a harvester for [Socrata Open Data API discovery
endpoints](https://socratadiscovery.docs.apiary.io/#), such as
the one offered by [Open Data Camden](https://opendata.camden.gov.uk).

It can be configured on the new harvester form by:
- Entering a URL - so far I have only tried https://opendata.camden.gov.uk
as that's all we need for this ticket
- Selecting SODA as the source type
- Adding 'app_token' to the configuration, with an application token
  associated with your account on the source site. We have one of these
for production use (see Sven for info), but for test/development it is
free and easy to create one yourself.

Socrata APIs don't have a concept of "dataset" vs "resource" - the
endpoint provides a list of resources, which may have relationships to
other resources. I have ignored these relationships and created one dataset
per resource for now

Files are not downloaded and stored in CKAN, as this would just be a
waste of space and another thing to worry about synchronising - we
just link to the resource on the source.

Not all resources in the Camden source catalogue have a link to an actual
file associated with them. Additionally, some have links that 404 -
another GIGO situation. I have decided to treat both in the same way: if
we can't get a 200 response from the resource download endpoint, we use
the link to the HTML page on the source website where users can view all
available information.

There's no guaruntee that this will work for other Socrata APIs, as not
all of them will have a web interface that works like this, but it works
for the Camden Open Data site and we can cross other bridges when we get to them.

The organisations are given only by name in the source API, so we
generate IDs for them based on the name. This means that we see
different organisations for different spellings of the same thing, such
as "London Borough of Camden Council" and "London Borough of Camden
Counil". This is a garbage-in, garbage-out situation, about which I am
not inclined to do anything. Some source datasets do not have an
associated organisation and will be attributed to whatever organisation
was given as default in the harvester. If the organisations generated
from this system are too messy, we can just omit the 'remote_orgs' key
from the config and use a default organisation for all of them.

Commit:
Add Redbridge harvester, update utils

Commit:
Add some comments

madelinekosse committed:
DAT-409 - Delete datasets no longer present in source

Check all datasets previously imported by this harvester and compare the IDs to
the list returned from the source so we can delete those that are no longer
present. Since the harvester is now aware of the datasets currently in CKAN, I
have used this information to check which ones are new so we only check for
changes when a dataset already exists and only add the organisation and extras
when it is new.

madelinekosse reviewed (commented):
Code looks good, but I haven't run the whole thing yet as I'm not sure how to configure it. Are you able to add a pre-harvest db dump as a reference for how the source should be configured?

Also I rebased my camden branch on this so I could use your util functions (they were useful :+1:  ) so will hold off on merging that one until this is done

madelinekosse reviewed (approved):
LGTM - ran it and it worked. Spat out some debug messages like this: `2023-08-11 12:07:47,543 DEBUG [ckanext.harvest.model] Failed to parse harvest object: 'action'` but everything worked as expected so it seems fine.

madelinekosse committed:
DAT-409 - Keep extras when updating dataset

When a dataset is updated by the harvester, merging the current and
harvested records overwrites the "extras" field, losing the data quality
and boost factor. This change adds the current values of these fields to
the record before saving the update by using the common util function. I
also refactored the part that adds default values to new dataset records
to use a common util function.

Commit:
Bugfix in Redbridge harvester, small update to utils

Commit:
Merge pull request #6 from GreaterLondonAuthority/feature/DAT-409-camden-harvest

DAT-409 - Add Socrata harvester
